File: 2015/3/3.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Location struct {
	X int
	Y int
}
func parseFile(filename string) []byte {
	file, err := os.Open(filename);
	if err != nil {
		fmt.Println("Error reading file.")
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var input []byte
	for scanner.Scan() {
		line := scanner.Text()
		input = []byte(line)
	}
	return input
}


func uniqueHouses(input []byte) int {
	locations := make(map[Location]struct{})
	x := 0
	y := 0
	initLocation := Location{X: x, Y: y}
	locations[initLocation] = struct{}{}
	for _, c := range input {
		switch c {
			case '>':
				x++
			case '<':
				x--
			case '^':
				y++
			default:
				y--
		} 
		if _, exists := locations[Location{X: x, Y: y}]; !exists {
			locations[Location{X: x, Y: y}] = struct{}{}
		}
	}
	return len(locations)
}

func uniqueHouses2(input []byte) int {
	locations := make(map[Location]struct{})
	x := 0
	y := 0
	a := 0
	b := 0
	initLocation := Location{X: x, Y: y}
	locations[initLocation] = struct{}{}
	for i, c := range input {
		if i % 2 == 0 {
			switch c {
				case '>':
					x++
				case '<':
					x--
				case '^':
					y++
				default:
					y--
			} 
			if _, exists := locations[Location{X: x, Y: y}]; !exists {
				locations[Location{X: x, Y: y}] = struct{}{}
			}
		} else {
			switch c {
				case '>':
					a--
				case '<':
					a++
				case '^':
					b--
				default:
					b++
			}
			if _, exists := locations[Location{X: a, Y: b}]; !exists {
				locations[Location{X: a, Y: b}] = struct{}{}
			}
		}
	}
	return len(locations)
}

func main() {
	filename := "3.txt"
	input := parseFile(filename)
	ans1 := uniqueHouses(input)
	ans2 := uniqueHouses2(input)
	fmt.Println("Answer for the part 1:", ans1)
	fmt.Println("Answer for the part 2:", ans2)
}

File: 2015/3/3.py
def parse_file(filename):
    parts = []
    with open(filename, 'r') as file:
        for line in file:
            parts = list(line.strip())
    return parts

def unique_houses(input):
    locations = set()
    x , y = 0, 0
    locations.add((x, y))
    for c in input:
        if c == '>':
            x += 1
        elif c == '<':
            x -= 1
        elif c == '^':
            y += 1
        else:
            y -= 1
        locations.add((x, y))
    return len(locations)

def unique_houses2(input):
    locations = set()
    x, y = 0, 0
    a, b = 0, 0
    locations.add((x, y))
    i = 0
    for c in input:
        if i % 2 == 0:
            if c == '>':
                x += 1
            elif c == '<':
                x -= 1
            elif c == '^':
                y += 1
            else:
                y -= 1
            locations.add((x, y))
        else:
            if c == '>':
                a -= 1
            elif c == '<':
                a += 1
            elif c == '^':
                b -= 1
            else:
                b += 1
            locations.add((a, b))
        i += 1
    return len(locations)


filename = "3.txt"
input = parse_file(filename)
ans1 = unique_houses(input)
ans2 = unique_houses2(input)
print("Answer for the part 1:", ans1)
print("Answer for the part 2:", ans2)
